README.md

```diff
@@ -1,11 +1,458 @@
+  - have unpredictable results, including data loss and security vulnerability.
+  - create an infinite regression of self-checking, because in theory
+    they could be buggy themselves.
+- Bugs go undetected so code quality suffers.
+- Even if bugs are logged:
+  - they are often deprioritized because they are not crashes, so
+    code quality suffers.
+  - developers are robbed of any chance to get a debuggable program
+    image; debugging becomes much harder, so code quality suffers.
+- Nobody can actually think through the implications of everything
+  being potentially buggy, so code beomes hard to reason about.
+  Maintenance is more error-prone and quality suffers.
+
+Note that in an unsafe language like C++, a seemingly recoverable
+condition like the discovery of a negative index can easily be the
+result of undefined behavior that also scrambled memory or causes
+“impossible” execution.
+
+## Development
+
+The usual procedures for development with cmake apply.  One typical
+set of commands might be:
+
+```sh
+cmake -Wno-dev -S . -B ../build -GNinja        # configure
+cmake --build ../build                         # build/rebuild after changes
+ctest --output-on-failure --test-dir ../build  # test
+```
+
+## Reference
+
+### Configuration
+
+The behavior of this library is configured by preprocessor symbols.
+
+Instead of passing many `-D` options in your compiler command-line,
+you can put the definition of these symbols in a header file and
+simply define `ADOBE_CONTRACT_CHECKS_CONFIGURATION` as the `#include`
+argument to that file:
+
+```sh
+c++ -D \
+    -I .\
+    ADOBE_CONTRACT_CHECKS_CONFIGURATION="<myproject/adobe_contract_checks_config.hpp>" \
+    myproject_source.cpp -o myproject_executable
+```
+
+If you are using this project via CMake, defining the same symbol in
+CMake, on the command line or in your will cause all clients of this
+library in your build to use that configuration file.
+
+This library can only have one configuration in an executable, so the
+privilege of configuring it belongs to the executable.  In CMake,
+
+```cmake
+if(PROJECT_IS_TOP_LEVEL)
+  set(ADOBE_CONTRACT_CHECKS_CONFIGURATION "<myproject/adobe_contract_checks_config.hpp>")
+endif()
+```
+
+#### Preprocessor configuration symbols
+
+##### Contract handler definition
+
+- `ADOBE_CONTRACT_VIOLATED_INLINE_BODY`: if you want the contract
+  violation handler to be defined inline, make this symbol expand to
+  its body.  When the body is lighter-weight than a call to the
+  handler would be, an inline handler can limit the code generated at
+  the use site. For example,
+
+  ```c++
+  #define ADOBE_CONTRACT_VIOLATED_INLINE_BODY { ADOBE_BUILTIN_TRAP(); }
+  ```
+
+  You might use this option in release builds when a minimal handler
+  is required.  **Note:** Defining a more complex handler inline
+  usually will increase binary sizes and may hurt performance.
+
+- `ADOBE_CONTRACT_VIOLATED_RETURNS`: define this symbol if your
```

Comment:
Do we _ever_ want this option? I could see the `INLINE_BODY` returning an error from the enclosing function, but that would require a mandate on how errors are returned from all functions using the library. Do we have an example?

Reply:
I'm happy to remove this option, but I can imagine the same people that want an exception wanting to log the problem and continue, which after all is slightly better than what happens when you compile assertions out.  Your call.

Reply:
I'm good with dropping it.

Reply:
Done.